In [1]:
import findspark
findspark.init()

In [2]:
import pandas as pd
pd.set_option('display.max_colwidth', None)

In [4]:
import warnings
warnings.filterwarnings('ignore')

In [3]:
from pyspark.sql.session import SparkSession

spark = (SparkSession.builder
    .appName("MDA2-Group Assignment")
    .getOrCreate())


Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).


## Import Data

### Dataset Info

This step is only needed in our course environment; other Spark environments you might see out there might not need this statement.

Each quarter, we publish downloadable files of Capital Bikeshare trip data. The data includes:

- **Duration** – Duration of trip
- **Start Date** – Includes start date and time
- **End Date** – Includes end date and time
- **Start Station** – Includes starting station name and number
- **End Station** – Includes ending station name and number
- **Bike Number** – Includes ID number of bike used for the trip
- **Member Type** – Indicates whether user was a "registered" member (Annual Member, 30-Day Member, or Day Key Member) or a "casual" rider (Single Trip, 24-Hour Pass, 3-Day Pass, or 5-Day Pass)

This data has been processed to remove trips that are taken by staff as they service and inspect the system, trips that are taken to/from any of our “test” stations at our warehouses, and any trips lasting less than 60 seconds (potentially false starts or users trying to re-dock a bike to ensure it's secure).

**NOTE**: The 3-Day Membership replaced the 5-Day Membership in Fall 2011.

https://capitalbikeshare.com/system-data


**IMPORTANT** **this documentation only applies for the CSV files up to March, 2020 (2020 Q1). From April, 2020, onwards, the schema for the data changes to this form (adding new columns):**

- ride_id                object
- rideable_type          object
- started_at             object
- ended_at               object
- start_station_name     object
- start_station_id      float64
- end_station_name       object
- end_station_id        float64
- start_lat             float64
- start_lng             float64
- end_lat               float64
- end_lng               float64
- member_casual          object --> = Member Type

dtype: object



WE DECIDED TO ONLY WORK WITH THE DATA STARTING FROM APRIL 2020 -ONLY WITH THE CSV FILES THAT HAD THE NEWER SHCEMA.

In [4]:
'''
from pyspark.sql.types import StructType, StructField, StringType, TimestampType

# Define the schema explicitly
schema = StructType([
    StructField("Duration", StringType(), True),
    StructField("Start date", StringType(), True),
    StructField("End date", StringType(), True),
    StructField("Start station number", StringType(), True),
    StructField("Start station", StringType(), True),
    StructField("End station number", StringType(), True),
    StructField("End station", StringType(), True),
    StructField("Bike number", StringType(), True),
    StructField("Member type", StringType(), True)
])
'''

'\nfrom pyspark.sql.types import StructType, StructField, StringType, TimestampType\n\n# Define the schema explicitly\nschema = StructType([\n    StructField("Duration", StringType(), True),\n    StructField("Start date", StringType(), True),\n    StructField("End date", StringType(), True),\n    StructField("Start station number", StringType(), True),\n    StructField("Start station", StringType(), True),\n    StructField("End station number", StringType(), True),\n    StructField("End station", StringType(), True),\n    StructField("Bike number", StringType(), True),\n    StructField("Member type", StringType(), True)\n])\n'

In [5]:
from pyspark.sql.types import StructType, StructField, StringType, TimestampType, FloatType

# Define the schema explicitly
schema = StructType([
    StructField("ride_id", StringType(), True),
    StructField("rideable_type", StringType(), True),
    StructField("started_at", StringType(), True),
    StructField("ended_at", StringType(), True),
    StructField("start_station_name", StringType(), True),
    StructField("start_station_id", FloatType(), True),
    StructField("end_station_name", StringType(), True),
    StructField("end_station_id", FloatType(), True),
    StructField("start_lat", FloatType(), True),
    StructField("start_lng", FloatType(), True),
    StructField("end_lat", FloatType(), True),
    StructField("end_lng", FloatType(), True),
    StructField("member_casual", StringType(), True)
])


In [6]:
# Read a single CSV file to check schema
single_file_df = spark.read.schema(schema).csv("hdfs://localhost:9000/group_assignment_mda2/raw/capitalbikeshare/202202-capitalbikeshare-tripdata.csv")

# Show the DataFrame to verify the schema
single_file_df.printSchema()
single_file_df.show()


root
 |-- ride_id: string (nullable = true)
 |-- rideable_type: string (nullable = true)
 |-- started_at: string (nullable = true)
 |-- ended_at: string (nullable = true)
 |-- start_station_name: string (nullable = true)
 |-- start_station_id: float (nullable = true)
 |-- end_station_name: string (nullable = true)
 |-- end_station_id: float (nullable = true)
 |-- start_lat: float (nullable = true)
 |-- start_lng: float (nullable = true)
 |-- end_lat: float (nullable = true)
 |-- end_lng: float (nullable = true)
 |-- member_casual: string (nullable = true)



+----------------+-------------+-------------------+-------------------+--------------------+----------------+--------------------+--------------+---------+---------+---------+----------+-------------+
|         ride_id|rideable_type|         started_at|           ended_at|  start_station_name|start_station_id|    end_station_name|end_station_id|start_lat|start_lng|  end_lat|   end_lng|member_casual|
+----------------+-------------+-------------------+-------------------+--------------------+----------------+--------------------+--------------+---------+---------+---------+----------+-------------+
|         ride_id|rideable_type|         started_at|           ended_at|  start_station_name|            null|    end_station_name|          null|     null|     null|     null|      null|member_casual|
|6D3E254769125C2D| classic_bike|2022-02-11 14:46:17|2022-02-11 15:01:29|       5th & K St NW|         31600.0| 14th & Irving St NW|       31124.0| 38.90304|-77.01903| 38.92855| -77.03223|     

In [7]:
# Read all CSV files in the directory using the defined schema
capitalbikeshare_raw = spark.read.schema(schema).csv("hdfs://localhost:9000/group_assignment_mda2/raw/capitalbikeshare/*.csv")


In [8]:
# Cache the DataFrame if needed
capitalbikeshare_raw.cache()


DataFrame[ride_id: string, rideable_type: string, started_at: string, ended_at: string, start_station_name: string, start_station_id: float, end_station_name: string, end_station_id: float, start_lat: float, start_lng: float, end_lat: float, end_lng: float, member_casual: string]

In [ ]:
# Print the schema to verify
capitalbikeshare_raw.printSchema()

# Perform a count to ensure there are no read errors
print(f"Number of records: {capitalbikeshare_raw.count()}")


In [10]:
# Check for nulls or other anomalies in the 'started_at' and 'ended_at' columns as they have caused a lot of trouble
capitalbikeshare_raw.select("started_at", "ended_at").summary("count", "min", "max").show()


+-------+-------------------+-------------------+
|summary|         started_at|           ended_at|
+-------+-------------------+-------------------+
|  count|           11246826|           11246826|
|    min|2020-04-01 00:25:48|2020-04-01 00:27:59|
|    max|         started_at|           ended_at|
+-------+-------------------+-------------------+



In [161]:
# Only if I need to restart the process from the beginning...
'''
capitalbikeshare_raw = capitalbikeshare_raw_copy
'''

In [18]:
# Copy the original df into a 'safety' variable
'''
capitalbikeshare_raw_copy = capitalbikeshare_raw

capitalbikeshare_raw_copy.limit(5).toPandas()
'''

ride_id  rideable_type           started_at             ended_at  \
0           ride_id  rideable_type           started_at             ended_at   
1  D88F352076694FFC   classic_bike  2023-08-03 17:33:41  2023-08-03 18:03:07   
2  6B733406B1E19446   classic_bike  2023-08-16 08:19:26  2023-08-16 08:34:07   
3  56145CE76A7B5A66   classic_bike  2023-08-25 08:19:20  2023-08-25 08:37:11   
4  11B7E5D90AE9DBE5   classic_bike  2023-08-15 17:56:22  2023-08-15 18:13:58   

              start_station_name  start_station_id       end_station_name  \
0             start_station_name               NaN       end_station_name   
1  5th St & Massachusetts Ave NW           31265.0  20th & Columbia Rd NW   
2                  5th & K St NW           31600.0        23rd & E St NW    
3                  5th & K St NW           31600.0        23rd & E St NW    
4                23rd & E St NW            31260.0          5th & K St NW   

   end_station_id  start_lat  start_lng    end_lat    end_lng  member_casual  
0             NaN        NaN        NaN        NaN        NaN  member_casual  
1         31133.0  38.900928 -77.018677  38.918037 -77.045486         member  
2         31260.0  38.903042 -77.019028  38.896103 -77.049881         member  
3         31260.0  38.903042 -77.019028  38.896103 -77.049881         member  
4         31600.0  38.896103 -77.049881  38.903042 -77.019028         member

In [11]:
from pyspark.sql.functions import col, to_timestamp

# Check for entries that are non-date values
capitalbikeshare_raw.filter(~col("started_at").rlike("^\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2}$")).select("started_at").show(truncate=False)


+----------+
|started_at|
+----------+
|started_at|
|started_at|
|started_at|
|started_at|
|started_at|
|started_at|
|started_at|
|started_at|
|started_at|
|started_at|
|started_at|
|started_at|
|started_at|
|started_at|
|started_at|
|started_at|
|started_at|
|started_at|
|started_at|
|started_at|
+----------+
only showing top 20 rows



In [11]:
from pyspark.sql.functions import expr


# Get the first row of the DataFrame
header = capitalbikeshare_raw.first()

# Get the name of the first column
first_column_name = capitalbikeshare_raw.columns[0]

# Filter out the rows that have the same value as the header in the first row
capitalbikeshare_filtered = capitalbikeshare_raw.filter(expr("NOT {} = '{}'".format(first_column_name, header[0])))

# Count the number of rows that were dropped or filtered
num_dropped = capitalbikeshare_raw.count() - capitalbikeshare_filtered.count()

# Print the number of rows that were dropped or filtered and the new total number of rows in the DataFrame
print(f"Dropped {num_dropped} rows. New total number of rows: {capitalbikeshare_filtered.count()}")



Dropped 42 rows. New total number of rows: 11246784


In [12]:
capitalbikeshare_raw.unpersist()
capitalbikeshare_filtered.cache()

DataFrame[ride_id: string, rideable_type: string, started_at: string, ended_at: string, start_station_name: string, start_station_id: float, end_station_name: string, end_station_id: float, start_lat: float, start_lng: float, end_lat: float, end_lng: float, member_casual: string]

In [13]:
capitalbikeshare_filtered.limit(5).toPandas()

ride_id rideable_type           started_at             ended_at  \
0  D88F352076694FFC  classic_bike  2023-08-03 17:33:41  2023-08-03 18:03:07   
1  6B733406B1E19446  classic_bike  2023-08-16 08:19:26  2023-08-16 08:34:07   
2  56145CE76A7B5A66  classic_bike  2023-08-25 08:19:20  2023-08-25 08:37:11   
3  11B7E5D90AE9DBE5  classic_bike  2023-08-15 17:56:22  2023-08-15 18:13:58   
4  D934138EBBF5BE96  classic_bike  2023-08-12 15:38:38  2023-08-12 15:46:22   

              start_station_name  start_station_id       end_station_name  \
0  5th St & Massachusetts Ave NW           31265.0  20th & Columbia Rd NW   
1                  5th & K St NW           31600.0        23rd & E St NW    
2                  5th & K St NW           31600.0        23rd & E St NW    
3                23rd & E St NW            31260.0          5th & K St NW   
4                  5th & K St NW           31600.0         11th & O St NW   

   end_station_id  start_lat  start_lng    end_lat    end_lng member_casual  
0         31133.0  38.900928 -77.018677  38.918037 -77.045486        member  
1         31260.0  38.903042 -77.019028  38.896103 -77.049881        member  
2         31260.0  38.903042 -77.019028  38.896103 -77.049881        member  
3         31600.0  38.896103 -77.049881  38.903042 -77.019028        member  
4         31286.0  38.903042 -77.019028  38.908432 -77.027092        member

In [25]:
# Check for entries that are non-date values
capitalbikeshare_filtered.filter(~col("started_at").rlike("^\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2}$")).select("started_at").show(truncate=False)


Exception in thread "serve RDD 141" java.net.SocketTimeoutException: Accept timed out
	at java.base/java.net.PlainSocketImpl.socketAccept(Native Method)
	at java.base/java.net.AbstractPlainSocketImpl.accept(AbstractPlainSocketImpl.java:458)
	at java.base/java.net.ServerSocket.implAccept(ServerSocket.java:565)
	at java.base/java.net.ServerSocket.accept(ServerSocket.java:533)
	at org.apache.spark.security.SocketAuthServer$$anon$1.run(SocketAuthServer.scala:64)


+----------+
|started_at|
+----------+
+----------+



In [14]:
capitalbikeshare_filtered.printSchema()

root
 |-- ride_id: string (nullable = true)
 |-- rideable_type: string (nullable = true)
 |-- started_at: string (nullable = true)
 |-- ended_at: string (nullable = true)
 |-- start_station_name: string (nullable = true)
 |-- start_station_id: float (nullable = true)
 |-- end_station_name: string (nullable = true)
 |-- end_station_id: float (nullable = true)
 |-- start_lat: float (nullable = true)
 |-- start_lng: float (nullable = true)
 |-- end_lat: float (nullable = true)
 |-- end_lng: float (nullable = true)
 |-- member_casual: string (nullable = true)



We need to modify the date columns to be operable timestamp columns

In [17]:
from pyspark.sql.functions import col, to_timestamp

# Rename the original date columns to indicate they are string representations
capitalbikeshare_dateOperable = capitalbikeshare_filtered.withColumnRenamed("started_at", "Start_date_str") \
                                            .withColumnRenamed("ended_at", "End_date_str")

# Convert the string date columns to timestamp data type with the original column names
capitalbikeshare_dateOperable = capitalbikeshare_dateOperable.withColumn(
    "started_at", to_timestamp(col("Start_date_str"), "yyyy-MM-dd HH:mm:ss")
).withColumn(
    "ended_at", to_timestamp(col("End_date_str"), "yyyy-MM-dd HH:mm:ss")
)

# Drop the original string date columns
capitalbikeshare_dateOperable = capitalbikeshare_dateOperable.drop("Start_date_str", "End_date_str")

# The DataFrame now has timestamp columns with the original names for date-based operations

In [18]:
capitalbikeshare_filtered.unpersist()
capitalbikeshare_dateOperable.cache()

DataFrame[ride_id: string, rideable_type: string, start_station_name: string, start_station_id: float, end_station_name: string, end_station_id: float, start_lat: float, start_lng: float, end_lat: float, end_lng: float, member_casual: string, started_at: timestamp, ended_at: timestamp]

In [19]:
capitalbikeshare_dateOperable.limit(5).toPandas()

/opt/spark3/python/pyspark/sql/pandas/conversion.py:194: FutureWarning: Passing unit-less datetime64 dtype to .astype is deprecated and will raise in a future version. Pass 'datetime64[ns]' instead
  series = series.astype(t, copy=False)
/opt/spark3/python/pyspark/sql/pandas/conversion.py:194: FutureWarning: Passing unit-less datetime64 dtype to .astype is deprecated and will raise in a future version. Pass 'datetime64[ns]' instead
  series = series.astype(t, copy=False)


ride_id rideable_type             start_station_name  \
0  D88F352076694FFC  classic_bike  5th St & Massachusetts Ave NW   
1  6B733406B1E19446  classic_bike                  5th & K St NW   
2  56145CE76A7B5A66  classic_bike                  5th & K St NW   
3  11B7E5D90AE9DBE5  classic_bike                23rd & E St NW    
4  D934138EBBF5BE96  classic_bike                  5th & K St NW   

   start_station_id       end_station_name  end_station_id  start_lat  \
0           31265.0  20th & Columbia Rd NW         31133.0  38.900928   
1           31600.0        23rd & E St NW          31260.0  38.903042   
2           31600.0        23rd & E St NW          31260.0  38.903042   
3           31260.0          5th & K St NW         31600.0  38.896103   
4           31600.0         11th & O St NW         31286.0  38.903042   

   start_lng    end_lat    end_lng member_casual          started_at  \
0 -77.018677  38.918037 -77.045486        member 2023-08-03 17:33:41   
1 -77.019028  38.896103 -77.049881        member 2023-08-16 08:19:26   
2 -77.019028  38.896103 -77.049881        member 2023-08-25 08:19:20   
3 -77.049881  38.903042 -77.019028        member 2023-08-15 17:56:22   
4 -77.019028  38.908432 -77.027092        member 2023-08-12 15:38:38   

             ended_at  
0 2023-08-03 18:03:07  
1 2023-08-16 08:34:07  
2 2023-08-25 08:37:11  
3 2023-08-15 18:13:58  
4 2023-08-12 15:46:22

In [30]:
# For each column, select the column, perform a count of nulls, and then show the results
for column in capitalbikeshare_dateOperable.columns:
    missing_count = capitalbikeshare_dateOperable.filter(col(column).isNull()).count()
    print(f"Column {column} has {missing_count} missing values")

Column ride_id has 0 missing values


Column rideable_type has 0 missing values


Column start_station_name has 680271 missing values


Column start_station_id has 680273 missing values


Column end_station_name has 753012 missing values


Column end_station_id has 753014 missing values


Column start_lat has 10 missing values


Column start_lng has 10 missing values


Column end_lat has 22184 missing values


Column end_lng has 22184 missing values


Column member_casual has 0 missing values


Column started_at has 0 missing values


Column ended_at has 0 missing values


In [33]:
# Find the minimum and maximum of the Start date (obviously it takes really long, 
# so I won't be using it after I've checked that it's correct).

from pyspark.sql.functions import min, max

# Find the minimum and maximum of the Start date
start_date_range = capitalbikeshare_dateOperable.agg(
    min("started_at").alias("Start date min"),
    max("started_at").alias("Start date max")
)

# Find the minimum and maximum of the End date
end_date_range = capitalbikeshare_dateOperable.agg(
    min("ended_at").alias("End date min"),
    max("ended_at").alias("End date max")
)

# Show the results
start_date_range.show()
end_date_range.show()

+-------------------+-------------------+
|     Start date min|     Start date max|
+-------------------+-------------------+
|2020-04-01 00:25:48|2023-09-30 23:59:58|
+-------------------+-------------------+



+-------------------+-------------------+
|       End date min|       End date max|
+-------------------+-------------------+
|2020-04-01 00:27:59|2023-10-05 02:00:19|
+-------------------+-------------------+



In [20]:
# Observe schema
capitalbikeshare_dateOperable.printSchema()

root
 |-- ride_id: string (nullable = true)
 |-- rideable_type: string (nullable = true)
 |-- start_station_name: string (nullable = true)
 |-- start_station_id: float (nullable = true)
 |-- end_station_name: string (nullable = true)
 |-- end_station_id: float (nullable = true)
 |-- start_lat: float (nullable = true)
 |-- start_lng: float (nullable = true)
 |-- end_lat: float (nullable = true)
 |-- end_lng: float (nullable = true)
 |-- member_casual: string (nullable = true)
 |-- started_at: timestamp (nullable = true)
 |-- ended_at: timestamp (nullable = true)



### Weather Dataset Info

https://www.visualcrossing.com/resources/documentation/weather-data/weather-data-documentation/

In [21]:
weatherdata_raw = (spark.read
                    .option("inferSchema", "true")
                    .option('header', 'true')
                    .csv("hdfs://localhost:9000/group_assignment_mda2/raw/weatherdata")
                    .cache())

In [22]:
weatherdata_raw.limit(5).toPandas()

/opt/spark3/python/pyspark/sql/pandas/conversion.py:194: FutureWarning: Passing unit-less datetime64 dtype to .astype is deprecated and will raise in a future version. Pass 'datetime64[ns]' instead
  series = series.astype(t, copy=False)
/opt/spark3/python/pyspark/sql/pandas/conversion.py:194: FutureWarning: Passing unit-less datetime64 dtype to .astype is deprecated and will raise in a future version. Pass 'datetime64[ns]' instead
  series = series.astype(t, copy=False)


name    datetime  tempmax  tempmin  temp  feelslikemax  \
0  washinton dc  2020-07-01     30.0     22.6  25.9          31.6   
1  washinton dc  2020-07-02     33.3     22.0  27.8          33.7   
2  washinton dc  2020-07-03     35.5     24.1  30.3          37.0   
3  washinton dc  2020-07-04     33.6     25.3  29.2          35.6   
4  washinton dc  2020-07-05     33.6     24.2  28.4          38.7   

   feelslikemin  feelslike   dew  humidity  ...  solarenergy  uvindex  \
0          22.6       26.5  20.0      70.6  ...          6.0        4   
1          22.0       28.1  18.0      57.9  ...          6.2        4   
2          24.1       31.0  18.0      49.2  ...          7.1        6   
3          25.3       30.7  20.0      58.3  ...          6.2        4   
4          24.2       30.7  22.4      71.3  ...          6.1        4   

   severerisk             sunrise              sunset  moonphase  \
0         NaN 2020-07-01 05:46:51 2020-07-01 20:37:16       0.37   
1         NaN 2020-07-02 05:47:21 2020-07-02 20:37:08       0.41   
2         NaN 2020-07-03 05:47:52 2020-07-03 20:36:58       0.45   
3         NaN 2020-07-04 05:48:24 2020-07-04 20:36:45       0.48   
4         NaN 2020-07-05 05:48:58 2020-07-05 20:36:31       0.50   

               conditions  \
0  Rain, Partially cloudy   
1        Partially cloudy   
2        Partially cloudy   
3  Rain, Partially cloudy   
4        Partially cloudy   

                                             description               icon  \
0            Partly cloudy throughout the day with rain.               rain   
1                      Becoming cloudy in the afternoon.  partly-cloudy-day   
2                      Partly cloudy throughout the day.  partly-cloudy-day   
3  Becoming cloudy in the afternoon with afternoon rain.               rain   
4                      Partly cloudy throughout the day.  partly-cloudy-day   

                                                                          stations  
0  KDCA,72405013743,72403793728,F0198,KGAI,KADW,KDAA,AS365,72033493764,74594013705  
1  KDCA,72405013743,72403793728,F0198,KGAI,KDAA,AS365,KADW,72033493764,74594013705  
2                         KDCA,72405013743,72403793728,F0198,KADW,KDAA,74594013705  
3                   KDCA,72405013743,72403793728,F0198,KADW,KDAA,AS365,74594013705  
4                         KDCA,72405013743,72403793728,F0198,KADW,KDAA,74594013705  

[5 rows x 33 columns]

In [37]:
# Observe schema
weatherdata_raw.printSchema()

root
 |-- name: string (nullable = true)
 |-- datetime: string (nullable = true)
 |-- tempmax: double (nullable = true)
 |-- tempmin: double (nullable = true)
 |-- temp: double (nullable = true)
 |-- feelslikemax: double (nullable = true)
 |-- feelslikemin: double (nullable = true)
 |-- feelslike: double (nullable = true)
 |-- dew: double (nullable = true)
 |-- humidity: double (nullable = true)
 |-- precip: double (nullable = true)
 |-- precipprob: integer (nullable = true)
 |-- precipcover: double (nullable = true)
 |-- preciptype: string (nullable = true)
 |-- snow: double (nullable = true)
 |-- snowdepth: double (nullable = true)
 |-- windgust: double (nullable = true)
 |-- windspeed: double (nullable = true)
 |-- winddir: double (nullable = true)
 |-- sealevelpressure: double (nullable = true)
 |-- cloudcover: double (nullable = true)
 |-- visibility: double (nullable = true)
 |-- solarradiation: double (nullable = true)
 |-- solarenergy: double (nullable = true)
 |-- uvindex: int

Analyzing columns PRELIMINARILY:

root
 |-- name: string (nullable = true) <--------------------- IRRELEVANT
 
 |-- datetime: string (nullable = true) <--------------------- NOT TIMESTAMP
 
 |-- tempmax: double (nullable = true) <--------------------- POSSIBLE REDUNDANCIES GROUP 1
 
 |-- tempmin: double (nullable = true) <--------------------- POSSIBLE REDUNDANCIES GROUP 1
 
 |-- temp: double (nullable = true) <--------------------- POSSIBLE REDUNDANCIES GROUP 1
 
 |-- feelslikemax: double (nullable = true) <--------------------- POSSIBLE REDUNDANCIES GROUP 2
 
 |-- feelslikemin: double (nullable = true) <--------------------- POSSIBLE REDUNDANCIES GROUP 2
 
 |-- feelslike: double (nullable = true) <--------------------- POSSIBLE REDUNDANCIES GROUP 2
 
 |-- dew: double (nullable = true)
 
 |-- humidity: double (nullable = true) <--------------------- POSSIBLE REDUNDANCIES GROUP 3
 
 |-- precip: double (nullable = true) <--------------------- POSSIBLE REDUNDANCIES GROUP 3
 
 |-- precipprob: integer (nullable = true) <--------------------- POSSIBLE REDUNDANCIES GROUP 3
 
 |-- precipcover: double (nullable = true) <--------------------- POSSIBLE REDUNDANCIES GROUP 3
 
 |-- preciptype: string (nullable = true)
 
 |-- snow: double (nullable = true)
 
 |-- snowdepth: double (nullable = true)
 
 |-- windgust: double (nullable = true) <--------------------- POSSIBLE REDUNDANCIES GROUP 4
 
 |-- windspeed: double (nullable = true) <--------------------- POSSIBLE REDUNDANCIES GROUP 4
 
 |-- winddir: double (nullable = true)
 
 |-- sealevelpressure: double (nullable = true)
 
 |-- cloudcover: double (nullable = true)
 
 |-- visibility: double (nullable = true)
 
 |-- solarradiation: double (nullable = true) <--------------------- POSSIBLE REDUNDANCIES GROUP 5
 
 |-- solarenergy: double (nullable = true) <--------------------- POSSIBLE REDUNDANCIES GROUP 5
 
 |-- uvindex: integer (nullable = true)
 
 |-- severerisk: integer (nullable = true)
 
 |-- sunrise: timestamp (nullable = true) <--------------------- EXTRACT TO 'sunlight_hours'
 
 |-- sunset: timestamp (nullable = true) <--------------------- EXTRACT TO 'sunlight_hours'
 
 |-- moonphase: double (nullable = true)
 
 |-- conditions: string (nullable = true) <--------------------- REDUNDANT & IMPRACTICAL
 
 |-- description: string (nullable = true) <--------------------- REDUNDANT & IMPRACTICAL
 
 |-- icon: string (nullable = true) <--------------------- REDUNDANT & IMPRACTICAL
 
 |-- stations: string (nullable = true) <--------------------- IRRELEVANT

In [63]:
# For the capitalbikeshare_raw DataFrame
capitalbikeshare_row_count = capitalbikeshare_dateOperable.count()
capitalbikeshare_column_count = len(capitalbikeshare_dateOperable.columns)
print(f"The shape of capitalbikeshare_dateOperable is: ({capitalbikeshare_row_count}, {capitalbikeshare_column_count})")

# For the weatherdata_raw DataFrame
weatherdata_row_count = weatherdata_raw.count()
weatherdata_column_count = len(weatherdata_raw.columns)
print(f"The shape of weatherdata_raw is: ({weatherdata_row_count}, {weatherdata_column_count})")


The shape of capitalbikeshare_dateOperable is: (11246784, 13)
The shape of weatherdata_raw is: (2133, 34)


#### Comment on Dataset Size

- **capitalbikeshare_raw** has 22,108,947 rows and 9 columns, which is quite sizable. Operations on this DataFrame, especially those that require a full pass over the data like .count(), .show(), or any kind of groupBy/aggregation, will take a noticeable amount of time to complete.
- **weatherdata_raw** has 2,133 rows and 33 columns, which is relatively small. Operations on this DataFrame should be much faster and are generally reasonable to perform without needing to sample or otherwise optimize for performance.

## Cleaning

In [23]:
from pyspark.sql.functions import to_timestamp

# Convert the 'datetime' column to timestamp data type
weatherdata_raw = weatherdata_raw.withColumn(
    "datetime", to_timestamp("datetime", "yyyy-MM-dd")
)

# Now the 'datetime' column is in timestamp format and ready for joining with the other dataset


weatherdata_raw.printSchema()

root
 |-- name: string (nullable = true)
 |-- datetime: timestamp (nullable = true)
 |-- tempmax: double (nullable = true)
 |-- tempmin: double (nullable = true)
 |-- temp: double (nullable = true)
 |-- feelslikemax: double (nullable = true)
 |-- feelslikemin: double (nullable = true)
 |-- feelslike: double (nullable = true)
 |-- dew: double (nullable = true)
 |-- humidity: double (nullable = true)
 |-- precip: double (nullable = true)
 |-- precipprob: integer (nullable = true)
 |-- precipcover: double (nullable = true)
 |-- preciptype: string (nullable = true)
 |-- snow: double (nullable = true)
 |-- snowdepth: double (nullable = true)
 |-- windgust: double (nullable = true)
 |-- windspeed: double (nullable = true)
 |-- winddir: double (nullable = true)
 |-- sealevelpressure: double (nullable = true)
 |-- cloudcover: double (nullable = true)
 |-- visibility: double (nullable = true)
 |-- solarradiation: double (nullable = true)
 |-- solarenergy: double (nullable = true)
 |-- uvindex: 

In [24]:
from pyspark.sql.functions import unix_timestamp, round

# Calculate 'sunlight_hours' by finding the difference between 'sunset' and 'sunrise', and convert to hours
weatherdata_raw = weatherdata_raw.withColumn(
    "sunlight_hours",
    round((unix_timestamp("sunset") - unix_timestamp("sunrise")) / 3600.0, 2)
)

# The 'sunlight_hours' column will now contain the duration of sunlight in hours, rounded to two decimal places


In [25]:
weatherdata_raw.limit(5).toPandas()

/opt/spark3/python/pyspark/sql/pandas/conversion.py:194: FutureWarning: Passing unit-less datetime64 dtype to .astype is deprecated and will raise in a future version. Pass 'datetime64[ns]' instead
  series = series.astype(t, copy=False)
/opt/spark3/python/pyspark/sql/pandas/conversion.py:194: FutureWarning: Passing unit-less datetime64 dtype to .astype is deprecated and will raise in a future version. Pass 'datetime64[ns]' instead
  series = series.astype(t, copy=False)
/opt/spark3/python/pyspark/sql/pandas/conversion.py:194: FutureWarning: Passing unit-less datetime64 dtype to .astype is deprecated and will raise in a future version. Pass 'datetime64[ns]' instead
  series = series.astype(t, copy=False)


name   datetime  tempmax  tempmin  temp  feelslikemax  \
0  washinton dc 2020-07-01     30.0     22.6  25.9          31.6   
1  washinton dc 2020-07-02     33.3     22.0  27.8          33.7   
2  washinton dc 2020-07-03     35.5     24.1  30.3          37.0   
3  washinton dc 2020-07-04     33.6     25.3  29.2          35.6   
4  washinton dc 2020-07-05     33.6     24.2  28.4          38.7   

   feelslikemin  feelslike   dew  humidity  ...  uvindex  severerisk  \
0          22.6       26.5  20.0      70.6  ...        4         NaN   
1          22.0       28.1  18.0      57.9  ...        4         NaN   
2          24.1       31.0  18.0      49.2  ...        6         NaN   
3          25.3       30.7  20.0      58.3  ...        4         NaN   
4          24.2       30.7  22.4      71.3  ...        4         NaN   

              sunrise              sunset  moonphase              conditions  \
0 2020-07-01 05:46:51 2020-07-01 20:37:16       0.37  Rain, Partially cloudy   
1 2020-07-02 05:47:21 2020-07-02 20:37:08       0.41        Partially cloudy   
2 2020-07-03 05:47:52 2020-07-03 20:36:58       0.45        Partially cloudy   
3 2020-07-04 05:48:24 2020-07-04 20:36:45       0.48  Rain, Partially cloudy   
4 2020-07-05 05:48:58 2020-07-05 20:36:31       0.50        Partially cloudy   

                                             description               icon  \
0            Partly cloudy throughout the day with rain.               rain   
1                      Becoming cloudy in the afternoon.  partly-cloudy-day   
2                      Partly cloudy throughout the day.  partly-cloudy-day   
3  Becoming cloudy in the afternoon with afternoon rain.               rain   
4                      Partly cloudy throughout the day.  partly-cloudy-day   

                                                                          stations  \
0  KDCA,72405013743,72403793728,F0198,KGAI,KADW,KDAA,AS365,72033493764,74594013705   
1  KDCA,72405013743,72403793728,F0198,KGAI,KDAA,AS365,KADW,72033493764,74594013705   
2                         KDCA,72405013743,72403793728,F0198,KADW,KDAA,74594013705   
3                   KDCA,72405013743,72403793728,F0198,KADW,KDAA,AS365,74594013705   
4                         KDCA,72405013743,72403793728,F0198,KADW,KDAA,74594013705   

   sunlight_hours  
0           14.84  
1           14.83  
2           14.82  
3           14.81  
4           14.79  

[5 rows x 34 columns]

In [42]:
# Obtaining all unique values in 'name' column (knowing they can't be many...), to ensure we should drop it.
unique_names = weatherdata_raw.select('name').distinct().collect()

# If there are only 1 or 2 unique values, this will be quick
unique_values = [row['name'] for row in unique_names]
unique_values


['washinton dc']

Explore missing values for weatherdata_raw:

In [43]:
from pyspark.sql.functions import count, when, isnan, lit
from pyspark.sql.types import DoubleType, FloatType

# Function to calculate missing values
def calculate_missing_values(df):
    total_rows = df.count()
    agg_exprs = []
    for c in df.columns:
        # For numerical columns, check for both null and NaN
        if isinstance(df.schema[c].dataType, (DoubleType, FloatType)):
            agg_exprs.append(count(when(col(c).isNull() | isnan(col(c)), c)).alias(c + '_count'))
            agg_exprs.append((count(when(col(c).isNull() | isnan(col(c)), c)) / total_rows * 100).alias(c + '_percent'))
        # For non-numerical columns, only check for null
        else:
            agg_exprs.append(count(when(col(c).isNull(), c)).alias(c + '_count'))
            agg_exprs.append((count(when(col(c).isNull(), c)) / total_rows * 100).alias(c + '_percent'))
    return df.agg(*agg_exprs)

# Use the function to calculate missing values on the DataFrame
missing_values_df = calculate_missing_values(weatherdata_raw)

# Convert to Pandas DataFrame for better formatting
missing_values_pandas = missing_values_df.toPandas()

# Transpose the DataFrame to have column names as the row index
missing_values_pandas = missing_values_pandas.transpose().reset_index()

# Split the 'index' column to separate the column names and the '_count' or '_percent' suffix
missing_values_pandas[['Column', 'Metric']] = missing_values_pandas['index'].str.rsplit('_', n=1, expand=True)

# Pivot the DataFrame to get the 'Missing Count' and 'Missing Percent' as separate columns
missing_values_pandas = missing_values_pandas.pivot(index='Column', columns='Metric', values=0)

# Reset the index to turn 'Column' back into a regular column
missing_values_pandas.reset_index(inplace=True)

# Rename the columns for clarity
missing_values_pandas.columns = ['Column', 'Missing Count', 'Missing Percent']

# Display the DataFrame
missing_values_pandas


Column  Missing Count  Missing Percent
0         cloudcover            0.0         0.000000
1         conditions            0.0         0.000000
2           datetime            0.0         0.000000
3        description            0.0         0.000000
4                dew            0.0         0.000000
5          feelslike            0.0         0.000000
6       feelslikemax            0.0         0.000000
7       feelslikemin            0.0         0.000000
8           humidity            0.0         0.000000
9               icon            0.0         0.000000
10         moonphase            0.0         0.000000
11              name            0.0         0.000000
12            precip            0.0         0.000000
13       precipcover            0.0         0.000000
14        precipprob            0.0         0.000000
15        preciptype         1144.0        53.633380
16  sealevelpressure            0.0         0.000000
17        severerisk         1470.0        68.917018
18              snow            0.0         0.000000
19         snowdepth            0.0         0.000000
20       solarenergy            0.0         0.000000
21    solarradiation            0.0         0.000000
22          stations            0.0         0.000000
23    sunlight_hours            0.0         0.000000
24           sunrise            0.0         0.000000
25            sunset            0.0         0.000000
26              temp            0.0         0.000000
27           tempmax            0.0         0.000000
28           tempmin            0.0         0.000000
29           uvindex            0.0         0.000000
30        visibility            0.0         0.000000
31           winddir            0.0         0.000000
32          windgust          555.0        26.019691
33         windspeed            0.0         0.000000

Drop the irrelevant columns and those with too many missing values (taking in to account the amount of other related weather variables):


In [26]:

weatherdata_clean = weatherdata_raw.drop('name', 'conditions', 'description', 'icon','preciptype','severerisk','windgust','stations')


In [27]:
weatherdata_clean.limit(5).toPandas()

/opt/spark3/python/pyspark/sql/pandas/conversion.py:194: FutureWarning: Passing unit-less datetime64 dtype to .astype is deprecated and will raise in a future version. Pass 'datetime64[ns]' instead
  series = series.astype(t, copy=False)
/opt/spark3/python/pyspark/sql/pandas/conversion.py:194: FutureWarning: Passing unit-less datetime64 dtype to .astype is deprecated and will raise in a future version. Pass 'datetime64[ns]' instead
  series = series.astype(t, copy=False)
/opt/spark3/python/pyspark/sql/pandas/conversion.py:194: FutureWarning: Passing unit-less datetime64 dtype to .astype is deprecated and will raise in a future version. Pass 'datetime64[ns]' instead
  series = series.astype(t, copy=False)


datetime  tempmax  tempmin  temp  feelslikemax  feelslikemin  feelslike  \
0 2020-07-01     30.0     22.6  25.9          31.6          22.6       26.5   
1 2020-07-02     33.3     22.0  27.8          33.7          22.0       28.1   
2 2020-07-03     35.5     24.1  30.3          37.0          24.1       31.0   
3 2020-07-04     33.6     25.3  29.2          35.6          25.3       30.7   
4 2020-07-05     33.6     24.2  28.4          38.7          24.2       30.7   

    dew  humidity  precip  ...  sealevelpressure  cloudcover  visibility  \
0  20.0      70.6   3.133  ...            1011.5        54.0        16.0   
1  18.0      57.9   0.000  ...            1012.0        36.9        16.0   
2  18.0      49.2   0.000  ...            1011.7        20.4        16.0   
3  20.0      58.3   0.537  ...            1012.7        40.9        15.7   
4  22.4      71.3   0.000  ...            1014.8        44.3        13.5   

   solarradiation  solarenergy  uvindex             sunrise  \
0            71.3          6.0        4 2020-07-01 05:46:51   
1            73.5          6.2        4 2020-07-02 05:47:21   
2            84.0          7.1        6 2020-07-03 05:47:52   
3            73.0          6.2        4 2020-07-04 05:48:24   
4            69.9          6.1        4 2020-07-05 05:48:58   

               sunset  moonphase  sunlight_hours  
0 2020-07-01 20:37:16       0.37           14.84  
1 2020-07-02 20:37:08       0.41           14.83  
2 2020-07-03 20:36:58       0.45           14.82  
3 2020-07-04 20:36:45       0.48           14.81  
4 2020-07-05 20:36:31       0.50           14.79  

[5 rows x 26 columns]

Verify no missing values for cleaned dataframe

In [46]:
# Use the function to verify missing values on the cleaned DataFrame ('wd_c' stands for 'weatherdata_clean')
missing_values_wd_c = calculate_missing_values(weatherdata_clean)

# Convert to Pandas DataFrame for better formatting
missing_values_wd_c_pandas = missing_values_wd_c.toPandas()

# Transpose the DataFrame to have column names as the row index
missing_values_wd_c_pandas = missing_values_wd_c_pandas.transpose().reset_index()

# Split the 'index' column to separate the column names and the '_count' or '_percent' suffix
missing_values_wd_c_pandas[['Column', 'Metric']] = missing_values_wd_c_pandas['index'].str.rsplit('_', n=1, expand=True)

# Pivot the DataFrame to get the 'Missing Count' and 'Missing Percent' as separate columns
missing_values_wd_c_pandas = missing_values_wd_c_pandas.pivot(index='Column', columns='Metric', values=0)

# Reset the index to turn 'Column' back into a regular column
missing_values_wd_c_pandas.reset_index(inplace=True)

# Rename the columns for clarity
missing_values_wd_c_pandas.columns = ['Column', 'Missing Count', 'Missing Percent']

# Display the DataFrame
missing_values_wd_c_pandas

Column  Missing Count  Missing Percent
0         cloudcover            0.0              0.0
1           datetime            0.0              0.0
2                dew            0.0              0.0
3          feelslike            0.0              0.0
4       feelslikemax            0.0              0.0
5       feelslikemin            0.0              0.0
6           humidity            0.0              0.0
7          moonphase            0.0              0.0
8             precip            0.0              0.0
9        precipcover            0.0              0.0
10        precipprob            0.0              0.0
11  sealevelpressure            0.0              0.0
12              snow            0.0              0.0
13         snowdepth            0.0              0.0
14       solarenergy            0.0              0.0
15    solarradiation            0.0              0.0
16    sunlight_hours            0.0              0.0
17           sunrise            0.0              0.0
18            sunset            0.0              0.0
19              temp            0.0              0.0
20           tempmax            0.0              0.0
21           tempmin            0.0              0.0
22           uvindex            0.0              0.0
23        visibility            0.0              0.0
24           winddir            0.0              0.0
25         windspeed            0.0              0.0

Explore missing values for capitalbikeshare_dateOperable:

In [ ]:
from pyspark.sql.types import DoubleType, FloatType
from pyspark.sql.functions import isnan, when

# Initialize Spark session
spark = SparkSession.builder.appName("missing_values").getOrCreate()

# Function to calculate missing values for each column
def calculate_missing_values(df):
    total_rows = df.count()
    agg_exprs_counts = [
        count(when(col(c).isNull(), c)).alias(c + '_count')
        for c in df.columns
    ]
    agg_exprs_percent = [
        (count(when(col(c).isNull(), c)) / total_rows * 100).alias(c + '_percent')
        for c in df.columns
    ]
    return df.agg(*agg_exprs_counts + agg_exprs_percent)

# Use the function to calculate missing values on the DataFrame
missing_values_cbs_r = calculate_missing_values(capitalbikeshare_dateOperable)

# Convert to Pandas DataFrame for better formatting
missing_values_cbs_r_pandas = missing_values_cbs_r.toPandas()

# Prepare the data for the desired output format
missing_counts_cbs_r_final = pd.DataFrame({
    'Column': [c.replace('_count', '') for c in missing_values_cbs_r_pandas.columns[:len(capitalbikeshare_dateOperable.columns)]],
    'Missing Count': missing_values_cbs_r_pandas.iloc[0, :len(capitalbikeshare_dateOperable.columns)].astype(int),
    'Missing Percent': missing_values_cbs_r_pandas.iloc[0, len(capitalbikeshare_dateOperable.columns):].values
})

# Format the 'Missing Percent' column to show as a percentage with two decimal places
missing_counts_cbs_r_final['Missing Percent'] = missing_counts_cbs_r_final['Missing Percent'].map(lambda x: '{:.2f}%'.format(x))

missing_counts_cbs_r_final



Drop irrelevant columns:

In [29]:
capitalbikeshare_dropped_cols = capitalbikeshare_dateOperable.drop('start_station_name', 'end_station_name', 'start_lat', 'start_lng', 'end_lat', 'end_lng')

In [31]:
capitalbikeshare_dateOperable.unpersist()

DataFrame[ride_id: string, rideable_type: string, start_station_name: string, start_station_id: float, end_station_name: string, end_station_id: float, start_lat: float, start_lng: float, end_lat: float, end_lng: float, member_casual: string, started_at: timestamp, ended_at: timestamp]

In [30]:
capitalbikeshare_dropped_cols.printSchema()

root
 |-- ride_id: string (nullable = true)
 |-- rideable_type: string (nullable = true)
 |-- start_station_id: float (nullable = true)
 |-- end_station_id: float (nullable = true)
 |-- member_casual: string (nullable = true)
 |-- started_at: timestamp (nullable = true)
 |-- ended_at: timestamp (nullable = true)



Drop all rows that contain missing values, as they are relatively few

In [32]:
# Drop all rows that contain missing values from the 'capitalbikeshare_dropped_cols' DataFrame
capitalbikeshare_clean = capitalbikeshare_dropped_cols.dropna()

# Count the number of rows in both DataFrames
rows_cbs_dateOperable_count = capitalbikeshare_dropped_cols.count()
rows_cbs_clean_count = capitalbikeshare_clean.count()

# Calculate the number and percentage of rows that were dropped
num_rows_dropped = rows_cbs_dateOperable_count - rows_cbs_clean_count
percent_dropped = (float(num_rows_dropped) / rows_cbs_dateOperable_count) * 100

# Print the number and percentage of rows that were dropped
print(num_rows_dropped)
print(percent_dropped)




1031804
9.174213713004535


Verify no more missing values:

In [70]:
# Use the function to calculate missing values on the DataFrame
missing_values_cbs_c = calculate_missing_values(capitalbikeshare_clean)

# Convert to Pandas DataFrame for better formatting
missing_values_cbs_c_pandas = missing_values_cbs_c.toPandas()

# Prepare the data for the desired output format
missing_counts_cbs_c_final = pd.DataFrame({
    'Column': [c.replace('_count', '') for c in missing_values_cbs_c_pandas.columns[:len(capitalbikeshare_clean.columns)]],
    'Missing Count': missing_values_cbs_c_pandas.iloc[0, :len(capitalbikeshare_clean.columns)].astype(int),
    'Missing Percent': missing_values_cbs_c_pandas.iloc[0, len(capitalbikeshare_clean.columns):].values
})

# Format the 'Missing Percent' column to show as a percentage with two decimal places
missing_counts_cbs_c_final['Missing Percent'] = missing_counts_cbs_c_final['Missing Percent'].map(lambda x: '{:.2f}%'.format(x))

missing_counts_cbs_c_final

Column  Missing Count Missing Percent
ride_id_count                    ride_id              0           0.00%
rideable_type_count        rideable_type              0           0.00%
start_station_id_count  start_station_id              0           0.00%
end_station_id_count      end_station_id              0           0.00%
member_casual_count        member_casual              0           0.00%
started_at_count              started_at              0           0.00%
ended_at_count                  ended_at              0           0.00%

## Joining Both Tables

In [33]:
from pyspark.sql.functions import to_date

# Convert the 'started_at' column to a date data type and store the result in a new column called 'start_date'
capitalbikeshare_clean = capitalbikeshare_clean.withColumn(
    "start_date", to_date("started_at")
)

# Import the unix_timestamp function from the pyspark.sql.functions module
from pyspark.sql.functions import unix_timestamp

# Calculate the duration of each bike ride by finding the difference between the 'ended_at' and 'started_at' columns,
# and store the result in a new column called 'duration'
capitalbikeshare_clean = capitalbikeshare_clean.withColumn(
    "duration", unix_timestamp("ended_at") - unix_timestamp("started_at")
)

In [34]:
# Move the 'start_date' column to the beginning of the 'capitalbikeshare_clean' DataFrame
capitalbikeshare_clean = capitalbikeshare_clean.select('start_date', *capitalbikeshare_clean.drop('start_date').columns)


In [35]:
capitalbikeshare_clean.cache()
capitalbikeshare_clean.limit(5).toPandas()

/opt/spark3/python/pyspark/sql/pandas/conversion.py:194: FutureWarning: Passing unit-less datetime64 dtype to .astype is deprecated and will raise in a future version. Pass 'datetime64[ns]' instead
  series = series.astype(t, copy=False)
/opt/spark3/python/pyspark/sql/pandas/conversion.py:194: FutureWarning: Passing unit-less datetime64 dtype to .astype is deprecated and will raise in a future version. Pass 'datetime64[ns]' instead
  series = series.astype(t, copy=False)


start_date           ride_id rideable_type  start_station_id  \
0  2023-08-03  D88F352076694FFC  classic_bike           31265.0   
1  2023-08-16  6B733406B1E19446  classic_bike           31600.0   
2  2023-08-25  56145CE76A7B5A66  classic_bike           31600.0   
3  2023-08-15  11B7E5D90AE9DBE5  classic_bike           31260.0   
4  2023-08-12  D934138EBBF5BE96  classic_bike           31600.0   

   end_station_id member_casual          started_at            ended_at  \
0         31133.0        member 2023-08-03 17:33:41 2023-08-03 18:03:07   
1         31260.0        member 2023-08-16 08:19:26 2023-08-16 08:34:07   
2         31260.0        member 2023-08-25 08:19:20 2023-08-25 08:37:11   
3         31600.0        member 2023-08-15 17:56:22 2023-08-15 18:13:58   
4         31286.0        member 2023-08-12 15:38:38 2023-08-12 15:46:22   

   duration  
0      1766  
1       881  
2      1071  
3      1056  
4       464

In [36]:
weatherdata_clean.cache()
weatherdata_clean.limit(5).toPandas()   

/opt/spark3/python/pyspark/sql/pandas/conversion.py:194: FutureWarning: Passing unit-less datetime64 dtype to .astype is deprecated and will raise in a future version. Pass 'datetime64[ns]' instead
  series = series.astype(t, copy=False)
/opt/spark3/python/pyspark/sql/pandas/conversion.py:194: FutureWarning: Passing unit-less datetime64 dtype to .astype is deprecated and will raise in a future version. Pass 'datetime64[ns]' instead
  series = series.astype(t, copy=False)
/opt/spark3/python/pyspark/sql/pandas/conversion.py:194: FutureWarning: Passing unit-less datetime64 dtype to .astype is deprecated and will raise in a future version. Pass 'datetime64[ns]' instead
  series = series.astype(t, copy=False)


datetime  tempmax  tempmin  temp  feelslikemax  feelslikemin  feelslike  \
0 2020-07-01     30.0     22.6  25.9          31.6          22.6       26.5   
1 2020-07-02     33.3     22.0  27.8          33.7          22.0       28.1   
2 2020-07-03     35.5     24.1  30.3          37.0          24.1       31.0   
3 2020-07-04     33.6     25.3  29.2          35.6          25.3       30.7   
4 2020-07-05     33.6     24.2  28.4          38.7          24.2       30.7   

    dew  humidity  precip  ...  sealevelpressure  cloudcover  visibility  \
0  20.0      70.6   3.133  ...            1011.5        54.0        16.0   
1  18.0      57.9   0.000  ...            1012.0        36.9        16.0   
2  18.0      49.2   0.000  ...            1011.7        20.4        16.0   
3  20.0      58.3   0.537  ...            1012.7        40.9        15.7   
4  22.4      71.3   0.000  ...            1014.8        44.3        13.5   

   solarradiation  solarenergy  uvindex             sunrise  \
0            71.3          6.0        4 2020-07-01 05:46:51   
1            73.5          6.2        4 2020-07-02 05:47:21   
2            84.0          7.1        6 2020-07-03 05:47:52   
3            73.0          6.2        4 2020-07-04 05:48:24   
4            69.9          6.1        4 2020-07-05 05:48:58   

               sunset  moonphase  sunlight_hours  
0 2020-07-01 20:37:16       0.37           14.84  
1 2020-07-02 20:37:08       0.41           14.83  
2 2020-07-03 20:36:58       0.45           14.82  
3 2020-07-04 20:36:45       0.48           14.81  
4 2020-07-05 20:36:31       0.50           14.79  

[5 rows x 26 columns]

In [81]:
capitalbikeshare_clean.printSchema()
weatherdata_clean.printSchema()
# Print the final row count for each DataFrame
print(f"capitalbikeshare_clean row count: {capitalbikeshare_clean.count()}")
print(f"weatherdata_clean row count: {weatherdata_clean.count()}")


root
 |-- ride_id: string (nullable = true)
 |-- rideable_type: string (nullable = true)
 |-- start_station_id: float (nullable = true)
 |-- end_station_id: float (nullable = true)
 |-- member_casual: string (nullable = true)
 |-- started_at: timestamp (nullable = true)
 |-- ended_at: timestamp (nullable = true)

root
 |-- datetime: timestamp (nullable = true)
 |-- tempmax: double (nullable = true)
 |-- tempmin: double (nullable = true)
 |-- temp: double (nullable = true)
 |-- feelslikemax: double (nullable = true)
 |-- feelslikemin: double (nullable = true)
 |-- feelslike: double (nullable = true)
 |-- dew: double (nullable = true)
 |-- humidity: double (nullable = true)
 |-- precip: double (nullable = true)
 |-- precipprob: integer (nullable = true)
 |-- precipcover: double (nullable = true)
 |-- snow: double (nullable = true)
 |-- snowdepth: double (nullable = true)
 |-- windspeed: double (nullable = true)
 |-- winddir: double (nullable = true)
 |-- sealevelpressure: double (nullabl

capitalbikeshare_clean row count: 10214980
weatherdata_clean row count: 2133


In [41]:
from pyspark.sql.functions import col

# Join the DataFrames on the 'start_date' and 'datetime' columns
biketrips_weather_total = capitalbikeshare_clean.join(
    weatherdata_clean,
    capitalbikeshare_clean.start_date == weatherdata_clean.datetime,
    'left'
)

# Drop the 'datetime' column, since it is redundant with 'start_date'
biketrips_weather_total = biketrips_weather_total.drop('datetime')


In [42]:
biketrips_weather_total.limit(5).toPandas()

/opt/spark3/python/pyspark/sql/pandas/conversion.py:194: FutureWarning: Passing unit-less datetime64 dtype to .astype is deprecated and will raise in a future version. Pass 'datetime64[ns]' instead
  series = series.astype(t, copy=False)
/opt/spark3/python/pyspark/sql/pandas/conversion.py:194: FutureWarning: Passing unit-less datetime64 dtype to .astype is deprecated and will raise in a future version. Pass 'datetime64[ns]' instead
  series = series.astype(t, copy=False)
/opt/spark3/python/pyspark/sql/pandas/conversion.py:194: FutureWarning: Passing unit-less datetime64 dtype to .astype is deprecated and will raise in a future version. Pass 'datetime64[ns]' instead
  series = series.astype(t, copy=False)
/opt/spark3/python/pyspark/sql/pandas/conversion.py:194: FutureWarning: Passing unit-less datetime64 dtype to .astype is deprecated and will raise in a future version. Pass 'datetime64[ns]' instead
  series = series.astype(t, copy=False)


start_date           ride_id rideable_type  start_station_id  \
0  2023-08-03  D88F352076694FFC  classic_bike           31265.0   
1  2023-08-16  6B733406B1E19446  classic_bike           31600.0   
2  2023-08-25  56145CE76A7B5A66  classic_bike           31600.0   
3  2023-08-15  11B7E5D90AE9DBE5  classic_bike           31260.0   
4  2023-08-12  D934138EBBF5BE96  classic_bike           31600.0   

   end_station_id member_casual          started_at            ended_at  \
0         31133.0        member 2023-08-03 17:33:41 2023-08-03 18:03:07   
1         31260.0        member 2023-08-16 08:19:26 2023-08-16 08:34:07   
2         31260.0        member 2023-08-25 08:19:20 2023-08-25 08:37:11   
3         31600.0        member 2023-08-15 17:56:22 2023-08-15 18:13:58   
4         31286.0        member 2023-08-12 15:38:38 2023-08-12 15:46:22   

   duration  tempmax  ...  sealevelpressure  cloudcover  visibility  \
0      1766     26.0  ...            1018.9        85.1        15.0   
1       881     30.5  ...            1013.1        57.5        16.0   
2      1071     33.3  ...            1011.1        81.6        15.7   
3      1056     31.2  ...            1008.5        78.5        15.0   
4       464     32.7  ...            1011.8        64.4        16.0   

   solarradiation  solarenergy  uvindex             sunrise  \
0           109.0          9.4        8 2023-08-03 06:10:45   
1           217.2         18.8        9 2023-08-16 06:22:27   
2           188.7         16.5        9 2023-08-25 06:30:34   
3            87.0          7.5        8 2023-08-15 06:21:32   
4           165.6         14.3        9 2023-08-12 06:18:50   

               sunset  moonphase  sunlight_hours  
0 2023-08-03 20:17:27       0.56           14.11  
1 2023-08-16 20:01:52       0.00           13.66  
2 2023-08-25 19:49:21       0.32           13.31  
3 2023-08-15 20:03:11       0.97           13.69  
4 2023-08-12 20:07:01       0.87           13.80  

[5 rows x 34 columns]

In [43]:
capitalbikeshare_clean.unpersist()
weatherdata_clean.unpersist()

DataFrame[datetime: timestamp, tempmax: double, tempmin: double, temp: double, feelslikemax: double, feelslikemin: double, feelslike: double, dew: double, humidity: double, precip: double, precipprob: int, precipcover: double, snow: double, snowdepth: double, windspeed: double, winddir: double, sealevelpressure: double, cloudcover: double, visibility: double, solarradiation: double, solarenergy: double, uvindex: int, sunrise: timestamp, sunset: timestamp, moonphase: double, sunlight_hours: double]

In [44]:
# Cache the joined DataFrame for faster access
biketrips_weather_total.cache()

# Display the first few rows of the joined DataFrame
biketrips_weather_total.limit(5).toPandas()

/opt/spark3/python/pyspark/sql/pandas/conversion.py:194: FutureWarning: Passing unit-less datetime64 dtype to .astype is deprecated and will raise in a future version. Pass 'datetime64[ns]' instead
  series = series.astype(t, copy=False)
/opt/spark3/python/pyspark/sql/pandas/conversion.py:194: FutureWarning: Passing unit-less datetime64 dtype to .astype is deprecated and will raise in a future version. Pass 'datetime64[ns]' instead
  series = series.astype(t, copy=False)
/opt/spark3/python/pyspark/sql/pandas/conversion.py:194: FutureWarning: Passing unit-less datetime64 dtype to .astype is deprecated and will raise in a future version. Pass 'datetime64[ns]' instead
  series = series.astype(t, copy=False)
/opt/spark3/python/pyspark/sql/pandas/conversion.py:194: FutureWarning: Passing unit-less datetime64 dtype to .astype is deprecated and will raise in a future version. Pass 'datetime64[ns]' instead
  series = series.astype(t, copy=False)


start_date           ride_id rideable_type  start_station_id  \
0  2023-08-03  D88F352076694FFC  classic_bike           31265.0   
1  2023-08-16  6B733406B1E19446  classic_bike           31600.0   
2  2023-08-25  56145CE76A7B5A66  classic_bike           31600.0   
3  2023-08-15  11B7E5D90AE9DBE5  classic_bike           31260.0   
4  2023-08-12  D934138EBBF5BE96  classic_bike           31600.0   

   end_station_id member_casual          started_at            ended_at  \
0         31133.0        member 2023-08-03 17:33:41 2023-08-03 18:03:07   
1         31260.0        member 2023-08-16 08:19:26 2023-08-16 08:34:07   
2         31260.0        member 2023-08-25 08:19:20 2023-08-25 08:37:11   
3         31600.0        member 2023-08-15 17:56:22 2023-08-15 18:13:58   
4         31286.0        member 2023-08-12 15:38:38 2023-08-12 15:46:22   

   duration  tempmax  ...  sealevelpressure  cloudcover  visibility  \
0      1766     26.0  ...            1018.9        85.1        15.0   
1       881     30.5  ...            1013.1        57.5        16.0   
2      1071     33.3  ...            1011.1        81.6        15.7   
3      1056     31.2  ...            1008.5        78.5        15.0   
4       464     32.7  ...            1011.8        64.4        16.0   

   solarradiation  solarenergy  uvindex             sunrise  \
0           109.0          9.4        8 2023-08-03 06:10:45   
1           217.2         18.8        9 2023-08-16 06:22:27   
2           188.7         16.5        9 2023-08-25 06:30:34   
3            87.0          7.5        8 2023-08-15 06:21:32   
4           165.6         14.3        9 2023-08-12 06:18:50   

               sunset  moonphase  sunlight_hours  
0 2023-08-03 20:17:27       0.56           14.11  
1 2023-08-16 20:01:52       0.00           13.66  
2 2023-08-25 19:49:21       0.32           13.31  
3 2023-08-15 20:03:11       0.97           13.69  
4 2023-08-12 20:07:01       0.87           13.80  

[5 rows x 34 columns]

In [45]:
from pyspark.sql.functions import col, sum

# Count the number of missing values in each column of the joined DataFrame
missing_values = biketrips_weather_total.select([sum(col(c).isNull().cast("int")).alias(c) for c in biketrips_weather_total.columns])

# Display the results
missing_values.show()


## Create Parquet File

In [ ]:
# Write the DataFrame back to HDFS in a processed format, like Parquet
biketrips_weather_total.write.parquet("hdfs://localhost:9000/group_assignment_mda2/processed/capitalbikeshare")

## End Spark Session

### Unpersist to free up resources

In [ ]:
biketrips_weather_total.unpersist()

### Tear Down

Once we complete the the lab we can stop all the services

### Stop Hadoop

Stops Hadoop
Open a terminal and execute
```sh
hadoop-stop.sh
```